pkg/ucp/aws/ucpcredentialprovider.go

```diff
@@ -73,19 +86,65 @@ func (c *UCPCredentialProvider) Retrieve(ctx context.Context) (aws.Credentials,
-		return aws.Credentials{}, errors.New("invalid access key info")
-	}
-
-	logger.Info(fmt.Sprintf("Retrieved AWS Credential - AccessKeyID: %s", s.AccessKeyCredential.AccessKeyID))
-
-	value := aws.Credentials{
-		AccessKeyID:     s.AccessKeyCredential.AccessKeyID,
-		SecretAccessKey: s.AccessKeyCredential.SecretAccessKey,
-		Source:          "radiusucp",
-		CanExpire:       true,
-		// Enables AWS SDK to fetch (rotate) access keys by calling Retrieve() after Expires.
-		Expires: time.Now().UTC().Add(c.options.Duration),
+	var value aws.Credentials
+	switch s.Kind {
+	case CredentialKindAccessKey:
+		if s.AccessKeyCredential == nil || s.AccessKeyCredential.AccessKeyID == "" || s.AccessKeyCredential.SecretAccessKey == "" {
+			return aws.Credentials{}, errors.New("invalid access key info")
+		}
+		logger.Info(fmt.Sprintf("Retrieved AWS Credential - AccessKeyID: %s", s.AccessKeyCredential.AccessKeyID))
+
+		value = aws.Credentials{
+			AccessKeyID:     s.AccessKeyCredential.AccessKeyID,
+			SecretAccessKey: s.AccessKeyCredential.SecretAccessKey,
+			Source:          "radiusucp",
+			CanExpire:       true,
+			Expires:         time.Now().UTC().Add(c.options.Duration),
+		}
+
+	case CredentialKindIRSA:
+		if s.IRSACredential == nil || s.IRSACredential.RoleARN == "" {
+			return aws.Credentials{}, errors.New("invalid IRSA info. RoleARN is required")
+		}
+		logger.Info(fmt.Sprintf("Retrieved AWS Credential - RoleARN: %s", s.IRSACredential.RoleARN))
+
+		// Radius requests will first be routed to STS endpoint,
+		// where it will be validated and then the request to the specific service (such as S3) will be made using
+		// the bearer token from the STS response.
+		// Based on the https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html,
+		// STS endpoint should be region based, and in the same region as
+		// Radius instance to minimize latency associated eith STS call and thereby improve performance.
+		// We should provide the user with ability to configure the STS endpoint region.
+		// For now, we are using the global STS endpoint, which is the default.
+		// Ref. https://github.com/radius-project/radius/issues/7747
+		awscfg, err := config.LoadDefaultConfig(context.TODO(),
+			config.WithRegion(awsSTSGlobalEndPointSigningRegion))
+
+		if err != nil {
+			return aws.Credentials{}, err
+		}
+
+		client := sts.NewFromConfig(awscfg)
+
+		credsCache := aws.NewCredentialsCache(stscreds.NewWebIdentityRoleProvider(
+			client,
+			s.IRSACredential.RoleARN,
+			stscreds.IdentityTokenFile(tokenFilePath),
+			func(o *stscreds.WebIdentityRoleOptions) {
+				o.RoleSessionName = "radius-ucp-" + uuid.New().String()
+			},
+		))
+
+		value, err = credsCache.Retrieve(ctx)
+		if err != nil {
+			logger.Info(fmt.Sprintf("Failed to retrieve AWS Credential IRSA - %s", err.Error()))
+			return aws.Credentials{}, err
+		}
+		value.Source = "radiusucp"
```

Comment:
Where is "radiusucp" coming from?

Reply:
I made it the same as the source we are setting for accesskey.. I understood it to be the source (holder) of credential and hence made it same.